question/bank/history/classes/question_history_view.php

```diff
@@ -0,0 +1,227 @@
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace qbank_history;
+
+use core_question\local\bank\view;
+
+/**
+ * Custom view class.
+ *
+ * @package    qbank_history
+ * @copyright  2021 Catalyst IT Australia Pty Ltd
+ * @author     Safat Shahin <safatshahin@catalyst-au.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class question_history_view extends view {
+
+    /**
+     * Entry id to get the versions
+     *
+     * @var int $entryid
+     */
+    protected $entryid;
+
+    /**
+     * Base url for the return.
+     *
+     * @var \moodle_url $basereturnurl
+     */
+    protected $basereturnurl;
+
+    /**
+     * Constructor.
+     * @param \core_question\lib\question_edit_contexts $contexts
+     * @param \moodle_url $pageurl
+     * @param \stdClass $course course settings
+     * @param \stdClass $cm activity settings.
+     * @param \stdClass $entryid quiz settings.
+     */
+    public function __construct($contexts, $pageurl, $course, $entryid, $returnurl, $cm = null) {
+        parent::__construct($contexts, $pageurl, $course, $cm);
+        $this->entryid = $entryid;
+        $this->basereturnurl = $returnurl;
+    }
+
+    protected function get_question_bank_plugins(): array {
+        $questionbankclasscolumns = [];
+        $corequestionbankcolumns = [
+                'checkbox_column',
+                'question_type_column',
+                'question_name_idnumber_tags_column',
+                'edit_menu_column',
+                'edit_action_column',
+                'tags_action_column',
+                'preview_action_column',
+                'delete_action_column',
+                'export_xml_action_column',
+                'question_status_column',
+                'version_number_column',
+                'creator_name_column',
+                'modifier_name_column',
+                'comment_count_column'
+        ];
+        if (question_get_display_preference('qbshowtext', 0, PARAM_BOOL, new \moodle_url(''))) {
+            $corequestionbankcolumns[] = 'question_text_row';
+        }
+
+        foreach ($corequestionbankcolumns as $fullname) {
+            $shortname = $fullname;
+            if (class_exists('qbank_history\\' . $fullname)) {
+                $fullname = 'qbank_history\\' . $fullname;
+                $questionbankclasscolumns[$shortname] = new $fullname($this);
+            } else if (class_exists('core_question\\local\\bank\\' . $fullname)) {
+                $fullname = 'core_question\\local\\bank\\' . $fullname;
+                $questionbankclasscolumns[$shortname] = new $fullname($this);
+            } else {
+                $questionbankclasscolumns[$shortname] = '';
+            }
+        }
+        $plugins = \core_component::get_plugin_list_with_class('qbank', 'plugin_feature', 'plugin_feature.php');
+        foreach ($plugins as $componentname => $plugin) {
```

Comment:
3 level deep for each here makes me twitch. I can't see much of a way around it.
You could extract a method here, but it might just be abstraction for the sake of it.

Reply:
because of having multiple plugins, and those plugins have multiple types of objects, and each type can pass multiple of those objects, it was a little tricky to make them come together. totally agree with this, I don't really like this multilevel loop. Kept it this way here as the base API implemented it this way, to keep some consistency.